File: regions/system/library/typescript/tracing/src/propagation.ts
import type { TraceContext } from './types.js';
import { toTraceparent, fromTraceparent } from './types.js';
import { Baggage } from './baggage.js';

export function injectContext(
  headers: Record<string, string>,
  ctx: TraceContext,
  baggage?: Baggage,
): void {
  headers['traceparent'] = toTraceparent(ctx);
  if (baggage) {
    const h = baggage.toHeader();
    if (h) {
      headers['baggage'] = h;
    }
  }
}

export function extractContext(headers: Record<string, string>): {
  context: TraceContext | null;
  baggage: Baggage;
} {
  const context = headers['traceparent'] ? fromTraceparent(headers['traceparent']) : null;
  const baggage = headers['baggage'] ? Baggage.fromHeader(headers['baggage']) : new Baggage();
  return { context, baggage };
}

File: regions/system/library/dart/tracing/lib/src/propagation.dart
import 'trace_context.dart';
import 'baggage.dart';

void injectContext(
  Map<String, String> headers,
  TraceContext ctx, [
  Baggage? baggage,
]) {
  headers['traceparent'] = ctx.toTraceparent();
  if (baggage != null) {
    final header = baggage.toHeader();
    if (header.isNotEmpty) {
      headers['baggage'] = header;
    }
  }
}

({TraceContext? context, Baggage baggage}) extractContext(
  Map<String, String> headers,
) {
  final traceparent = headers['traceparent'];
  final ctx = traceparent != null
      ? TraceContext.fromTraceparent(traceparent)
      : null;

  final baggageHeader = headers['baggage'];
  final baggage = baggageHeader != null
      ? Baggage.fromHeader(baggageHeader)
      : Baggage();

  return (context: ctx, baggage: baggage);
}
